velox/functions/prestosql/tests/RandTest.cpp

```diff
@@ -74,5 +93,146 @@ TEST_F(RandTest, nonNullInt8) {
+  EXPECT_LT(result, 23286);
+  EXPECT_GE(result, 0);
+}
+
+TEST_F(RandTest, secureRandInt8) {
+  auto result = secureRand<int8_t>((int8_t)10, (int8_t)120);
+  EXPECT_LT(result, 120);
+  EXPECT_GE(result, 10);
+
+  result = secureRand<int8_t>((int8_t)0, (int8_t)120);
+  EXPECT_LT(result, 120);
+  EXPECT_GE(result, 0);
+}
+
+TEST_F(RandTest, secureRandDouble) {
+  auto result = secureRand<double>((double)10.5, (double)120.7895);
+  EXPECT_LT(result, 120.7895);
+  EXPECT_GE(result, 10.5);
+
+  result = secureRand<double>((double)0.0, (double)120.7895);
+  EXPECT_LT(result, 120.7895);
+  EXPECT_GE(result, 0.0);
+}
+
+TEST_F(RandTest, secureRandFloat) {
+  auto result = secureRand<float>((float)-10.5, (float)120.7);
+  EXPECT_LT(result, 120.7);
+  EXPECT_GE(result, -10.5);
+
+  result = secureRand<float>((float)0.0, (float)120.7);
+  EXPECT_LT(result, 120.7);
+  EXPECT_GE(result, 0.0);
+}
+
+TEST_F(RandTest, secureRandInvalid) {
+  VELOX_ASSERT_THROW(
+      secureRand<int64_t>((int64_t)-5, (int64_t)-10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<int64_t>((int64_t)15, (int64_t)10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<int32_t>((int32_t)5, (int32_t)-10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<int32_t>((int32_t)15, (int32_t)10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<int16_t>((int16_t)-5, (int16_t)-10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<int16_t>((int16_t)15, (int16_t)10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<int8_t>((int8_t)5, (int8_t)-10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<int8_t>((int8_t)15, (int8_t)10),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<double>(-5.7, -10.7),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<double>(15.6, 10.1),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<float>((float)-5.7, (float)-10.7),
+      "upper bound must be greater than lower bound");
+  VELOX_ASSERT_THROW(
+      secureRand<float>((float)15.6, (float)10.1),
+      "upper bound must be greater than lower bound");
+}
+
+TEST_F(RandTest, secureRandSpecialValues) {
+  auto result = secureRand<int64_t>(0, kLongMax);
+  EXPECT_LT(result, kLongMax);
+  result = secureRand<int64_t>(kLongMin, 0);
+  EXPECT_GE(result, kLongMin);
+
+  result = secureRand<double>(-kInf, 0);
+  EXPECT_GE(result, -kInf);
+  result = secureRand<double>(0.0, kInf);
+  EXPECT_LE(result, kInf);
+
+  result = secureRand<float>(-kInfF, 0);
+  EXPECT_GE(result, -kInfF);
+  result = secureRand<float>(0.0, kInfF);
+  EXPECT_LE(result, kInfF);
```

Comment:
here result is of type std::optional<int64_t> so the result of secure_rand is getting cast to a int64_t which is doing weird things since the result is NaN and NaN can't be represented as an int64_t

(this was caught by UBSan)